CUDADataFormats/SiPixelDigi/src/SiPixelDigisCUDA.cc

```diff
@@ -1,28 +1,34 @@
+#include <cassert>
+
 #include "CUDADataFormats/SiPixelDigi/interface/SiPixelDigisCUDA.h"
 #include "HeterogeneousCore/CUDAUtilities/interface/copyAsync.h"
 #include "HeterogeneousCore/CUDAUtilities/interface/device_unique_ptr.h"
 #include "HeterogeneousCore/CUDAUtilities/interface/host_unique_ptr.h"
 
 SiPixelDigisCUDA::SiPixelDigisCUDA(size_t maxFedWords, cudaStream_t stream)
-    : m_store(cms::cuda::make_device_unique<uint16_t[]>(
-          maxFedWords * int(SiPixelDigisCUDASOAView::StorageLocation::MAX), stream)),
-      m_view(cms::cuda::make_host_unique<SiPixelDigisCUDASOAView>(stream)) {
-  auto get16 = [&](SiPixelDigisCUDASOAView::StorageLocation s) { return m_store.get() + int(s) * maxFedWords; };
-  m_view->xx_ = get16(SiPixelDigisCUDASOAView::StorageLocation::XX);
-  m_view->yy_ = get16(SiPixelDigisCUDASOAView::StorageLocation::YY);
-  m_view->adc_ = get16(SiPixelDigisCUDASOAView::StorageLocation::ADC);
-  m_view->moduleInd_ = get16(SiPixelDigisCUDASOAView::StorageLocation::MODULEIND);
-  m_view->clus_ = reinterpret_cast<int32_t*>(get16(SiPixelDigisCUDASOAView::StorageLocation::CLUS));
-  m_view->pdigi_ = reinterpret_cast<uint32_t*>(get16(SiPixelDigisCUDASOAView::StorageLocation::PDIGI));
-  m_view->rawIdArr_ = reinterpret_cast<uint32_t*>(get16(SiPixelDigisCUDASOAView::StorageLocation::RAWIDARR));
+    : m_store(cms::cuda::make_device_unique<SiPixelDigisCUDA::m_store_type[]>(
+          roundFor128ByteAlignment(maxFedWords) * int(SiPixelDigisCUDASOAView::StorageLocation::kMAX), stream)) {
+  assert(maxFedWords != 0);
+  auto get16 = [&](SiPixelDigisCUDASOAView::StorageLocation s) {
+    return m_store.get() + int(s) * roundFor128ByteAlignment(maxFedWords);
+  };
+  m_view.xx_ = get16(SiPixelDigisCUDASOAView::StorageLocation::kXX);
```

Comment:
`get16` could simply be a private member function ?

Reply:
it could also be used to simplify the interface:
```c++
template <typename T>
T* SiPixelDigisCUDA::getColumnAddress(SiPixelDigisCUDASOAView::StorageLocation column) {
    return reinterpret_cast<T*>(m_store.get() + static_cast<int>(column) * roundFor128ByteAlignment(maxFedWords));
}
```
to be used as
```c++
  ...
  m_view.moduleInd_{getColumnAddress<int16_t>(SiPixelDigisCUDASOAView::StorageLocation::kMODULEIND)},
  m_view.clus_{getColumnAddress<int32_t>(SiPixelDigisCUDASOAView::StorageLocation::kCLUS)},
  ...
```

Reply:
How about adding it to `SiPixelDigisCUDASOAView`? I was experimenting a bit with using it in the constructor.